### DISCUSSION TITLE
Simulating delay on specific Websocket messages

### DISCUSSION THREAD

**Discussion by renatosc** [mitmproxy/mitmproxy#5829] [Q&A]
Hi. I want to simulate a scenario where only specific websocket messages gets delayed (let's say they got a bad route to the destination which makes them arrive later).

Based on the [websocket-simple.py example](https://docs.mitmproxy.org/stable/addons-examples/#websocket-inject-message) I just replaced the 
`message.drop()`
with
`await asyncio.sleep(5)`
and added the required changes for async (import statement and async prefix on websocket_message method)

That code does create a delay on any FOOBAR message but it also delays any subsequent message, so let's say I send 3 messages one after the other right away:
**a**
**FOOBAR**
**b**

Message **a** is sent right away, **FOOBAR** is delayed but **b** is also delayed, arriving right after **FOOBAR**. So, the sleep appears to delay the entire queue not just a single message. I would like to see the message **b** arriving before FOOBAR at the server.

Is there anyway to achieve that?

**Comment by mhils** [comment]
> That code does create a delay on any FOOBAR message but it also delays any subsequent message

I think that is somewhat intentional behavior, we don't want mitmproxy to reorder messages when order is potentially important. I think the workaround for your case here is to call [`message.drop()`](https://docs.mitmproxy.org/dev/api/mitmproxy/websocket.html#WebSocketMessage.drop) and then [re-inject the message](https://github.com/mitmproxy/mitmproxy/blob/main/examples/addons/websocket-inject-message.py) later on from a new asyncio task (see second part of the example).

**Reply by renatosc** [reply]
Thanks for the reply @mhils. I agree that would be the normal behavior (and Fiddler works the same way also), but I really wanted to that specific test on the network level (and not simply simulate a delay in the server side code receiving the message).

I am actually trying to do that right now (dropping and injecting again)... I will update here when I am finished
